### ISSUE TITLE
How Can I get the matched nodes?

### ISSUE THREAD

**Issue by XanderPENG** [wannesm/LeuvenMapMatching#28]
I have tested several times that if I set "use_latlon=True" then the matching will be perfect, however, "matcher.path_pred_onlynodes" will return an empty list, which means that I cannot extract the matched points and output. 

What is interesting is that if I set "use_latlon=False", I can get the nodes list while the matching result is not so good.

I am wondering how this could happen and how to address it.

**Comment by debjitbhowmick** [comment]
Depends on the previous code on how you developed your map_con. Could you please share that?

**Comment by angelyix** [comment]
I had the same question, and it seems that the results of matcher.path_pred_onlynodes is not the matched points, instead of the starting nodes of the matched roads. I wonder how to get the matched points and are the points projected orthogonally onto the road network?

**Comment by XanderPENG** [comment]
> Depends on the previous code on how you developed your map_con. Could you please share that?

Yes, there was something wrong with the `map_con` setting in my memory. Since it has been a quite long time, I cannot find the previous codes. However, I still remember that I used the "osmnx" package to derive map directly and built the subsequent `map_con`, which works very well.

**Comment by XanderPENG** [comment]
> I had the same question, and it seems that the results of matcher.path_pred_onlynodes is not the matched points, instead of the starting nodes of the matched roads. I wonder how to get the matched points and are the points projected orthogonally onto the road network?

I had the same thoughts previously while I found that this requirement cannot be fulfilled by using this package only (Maybe could but I do not know how to achieve it). Thus, I firstly use this package to match the trajectory points into the road and get the instance of "matched road", then use the "Near" principle to match origin trajectory points with the matched road.

**Comment by wannesm** [comment]
Setting `use_latlon` should indeed have no direct influence on `matcher.path_pred_onlynodes`. It will return a different matching as you indicate but it should be the matching you see. What do you mean with a 'perfect matching'? Is it based on the visualization?

The `matcher.path_pred_onlynodes` method tries to simplify the matched path to only the nodes in the map that are passed through. The states return by `states, last_idx = matcher.match(track)` should also be informative.

You can also get access to the detailed mapping by looking at `matcher.lattice_best` (including where each matched point is on an edge etc). This is the datastructure that is also used for plotting. See this section in the docs: https://leuvenmapmatching.readthedocs.io/en/latest/usage/debug.html#inspect-the-best-matching .

**Comment by angelyix** [comment]
>I'm glad that my solution idea is quite similar to yours, I first applied this package to get the road geometry information, such as 'osmid' and length of every road segment. Then, I used 'Arcpy' package in Pycharm for nearest neighbor search to orthogonally project the trajectory points to the matched road. So what tools did you use to process the "Near" principle? : )

**Comment by angelyix** [comment]
> > I had the same question, and it seems that the results of matcher.path_pred_onlynodes is not the matched points, instead of the starting nodes of the matched roads. I wonder how to get the matched points and are the points projected orthogonally onto the road network?
> 
> I had the same thoughts previously while I found that this requirement cannot be fulfilled by using this package only (Maybe could but I do not know how to achieve it). Thus, I firstly use this package to match the trajectory points into the road and get the instance of "matched road", then use the "Near" principle to match origin trajectory points with the matched road.

I'm glad that my solution idea is quite similar to yours, I first applied this package to get the road geometry information, such as 'osmid' and length of every road segment. Then, I used 'Arcpy' package in Pycharm for nearest neighbor search to orthogonally project the trajectory points to the matched road. So what tools did you use to process the "Near" principle? : )

**Comment by wannesm** [comment]
To clarify, you do not necessarily need an additional package to find the nearest matching point since the matching is doing this already (otherwise a package like ArcPy is a good choice).

For example for the first matched point (thus with index 0), you can do:
```
match = matcher.lattice_best[0]
location_on_gps_trace = match.edge_o.pi
location_on_map = match.edge_m.pi
```

**Comment by debjitbhowmick** [comment]
> To clarify, you do not necessarily need an additional package to find the nearest matching point since the matching is doing this already (otherwise a package like ArcPy is a good choice).
> 
> For example for the first matched point (thus with index 0), you can do:
> 
> ```
> match = matcher.lattice_best[0]
> location_on_gps_trace = match.edge_o.pi
> location_on_map = match.edge_m.pi
> ```

Hi Wannes,

Thanks for your response. It really helped. I have a query regarding the **plotting of the matched points**.

It seems that the underlying code assumes that the matched OSM ways are straight lines, and therefore the matched GPS points seem to be not superimposed on the underlying road network when one single OSM way has some curvature. While this does not impact the map matching at all, but I would still like to get hold of the matched points on the road network. Before I dive into writing my own code for this, I was hoping if the it is possible to account for the curvature (inside the package itself)? I have attached some code and images for your reference. 

- Red points are observed GPS points, 
- Green points are the matched GPS points, and 
- Blue lines are the matched OSM ways. 
- To reiterate, the matching is perfect. 


![image](https://user-images.githubusercontent.com/42370063/193481947-c9405a28-2240-42a9-9363-dbf56662db1f.png)
![image](https://user-images.githubusercontent.com/42370063/193482033-07c31429-1d44-4837-ba21-c5b122720dcc.png)

The relevant code block is as follows:

![image](https://user-images.githubusercontent.com/42370063/193482077-d2dff1e5-4065-4758-98ee-5e75c9849235.png)
![image](https://user-images.githubusercontent.com/42370063/193482106-9dec3f52-e53e-4d46-aad0-19a9354bee2d.png)
![image](https://user-images.githubusercontent.com/42370063/193482152-63444b35-7e05-489c-8a47-9d970eeff365.png)
![image](https://user-images.githubusercontent.com/42370063/193482240-68834527-e6cb-44f8-8ebe-e5f9ff25c53c.png)
![image](https://user-images.githubusercontent.com/42370063/193482268-5f7bb398-88e9-4bc7-922d-ae8ae97b8029.png)
![image](https://user-images.githubusercontent.com/42370063/193482300-cc6c7a68-1b48-4494-823b-be4b56b0233a.png)

**Comment by debjitbhowmick** [comment]
Update: I employed the `nearest_points `function from `shapely.ops` to get the matched points on the road network. These were the corresponding results. I hope this helps others.

![image](https://user-images.githubusercontent.com/42370063/193483752-6db62396-0a17-4845-9d6e-014e8594ee78.png)
![image](https://user-images.githubusercontent.com/42370063/193483714-5db483b4-c9b4-41fd-888b-e6a7a6017706.png)

**Comment by wannesm** [comment]
If you observe straight lines, this is probably caused by how you generate the graph that is given to the toolbox. Assuming you also use osmnx, you have to tell osmnx not to simplify curved lines into straight lines (see also issue #34):

```
graph = ox.graph_from_place('Leuven, Belgium', network_type='drive', simplify=False)
```

**Comment by debjitbhowmick** [comment]
Thanks. That was really helpful. `Simplify=False` had them plotted on the edges of the graph. 

One more question. Does this affect the performance of map matching? I could not notice any change between the map matching performance while using the simplified vs un-simplified graphs? Only that the un-simplified graph map matching takes longer.
